crates/server/src/datastore/memory.rs

```diff
@@ -0,0 +1,491 @@
+        &self,
+    ) -> Result<
+        Pin<Box<dyn Stream<Item = Result<InitialLeaf, DataStoreError>> + Send>>,
+        DataStoreError,
+    > {
+        Ok(Box::pin(futures::stream::empty()))
+    }
+
+    async fn store_operator_record(
+        &self,
+        log_id: &LogId,
+        record_id: &RecordId,
+        record: &ProtoEnvelope<operator::OperatorRecord>,
+    ) -> Result<(), DataStoreError> {
+        let mut state = self.0.lock().await;
+        let prev = state.records.entry(log_id.clone()).or_default().insert(
+            record_id.clone(),
+            RecordStatus::Pending(PendingRecord::Operator {
+                record: Some(record.clone()),
+            }),
+        );
+
+        assert!(prev.is_none());
+        Ok(())
+    }
+
+    async fn reject_operator_record(
+        &self,
+        log_id: &LogId,
+        record_id: &RecordId,
+        reason: &str,
+    ) -> Result<(), DataStoreError> {
+        let mut state = self.0.lock().await;
+
+        match state
+            .records
+            .get_mut(log_id)
+            .ok_or_else(|| DataStoreError::LogNotFound(log_id.clone()))?
+            .get_mut(record_id)
+        {
+            Some(s @ RecordStatus::Pending(PendingRecord::Operator { .. })) => {
+                *s = RecordStatus::Rejected(reason.to_string());
+                Ok(())
+            }
+            _ => Err(DataStoreError::RecordNotFound(record_id.clone())),
+        }
+    }
+
+    async fn accept_operator_record(
+        &self,
+        log_id: &LogId,
+        record_id: &RecordId,
+    ) -> Result<(), DataStoreError> {
+        let mut state = self.0.lock().await;
+
+        let State {
+            operators, records, ..
+        } = &mut *state;
+
+        let status = records
+            .get_mut(log_id)
+            .ok_or_else(|| DataStoreError::LogNotFound(log_id.clone()))?
+            .get_mut(record_id)
+            .ok_or_else(|| DataStoreError::RecordNotFound(record_id.clone()))?;
+
+        let res = match status {
+            RecordStatus::Pending(PendingRecord::Operator { record }) => {
+                let record = record.take().unwrap();
+                let log = operators.entry(log_id.clone()).or_default();
+                let snapshot = log.validator.snapshot();
+
+                match unsafe { log.validator.validate_record(&record) } {
+                    Ok(()) => {
+                        let index = log.entries.len();
+                        log.entries.push(record);
+                        Ok(Record {
+                            index,
+                            checkpoint_index: None,
+                            sources: Default::default(),
+                        })
+                    }
+                    Err(e) => {
+                        log.validator.rollback(snapshot);
+                        Err(e.into())
+                    }
+                }
+            }
+            _ => Err(DataStoreError::RecordNotFound(record_id.clone())),
```

Comment:
`DataStoreError::RecordNotFound` seems to be returned for records that are already accepted or rejected.

Reply:
Added a `RecordNotPending` error variant for this case.